compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESinkIntoCore.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "aie/AIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-amdaie-sink-into-core"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+template <typename OpTy>
+class MoveAllDependenciesIntoOp : public OpRewritePattern<OpTy> {
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+  LogicalResult matchAndRewrite(OpTy parentOp,
+                                PatternRewriter &rewriter) const override {
+    bool changed = false;
+    SmallVector<Operation *> opsInParent;
+    parentOp->walk([&](Operation *op) {
+      if (op == parentOp) return WalkResult::advance();
+      opsInParent.push_back(op);
+      return WalkResult::advance();
+    });
+    for (auto opInParent : opsInParent) {
+      for (Value operand : opInParent->getOperands()) {
+        if (!operand || !operand.getDefiningOp()) continue;
+        Operation *dependencyOp = operand.getDefiningOp();
+        if (!parentOp->isProperAncestor(dependencyOp)) {
+          Region &r = parentOp->getRegion(0);
+          assert(r.getBlocks().size() == 1 && "expected single block region");
+          rewriter.setInsertionPointToStart(&r.front());
+          Operation *sunkOp = rewriter.clone(*dependencyOp);
+          dependencyOp->replaceUsesWithIf(sunkOp, [&](OpOperand &use) {
+            return parentOp->isProperAncestor(use.getOwner());
+          });
+          changed = true;
+        }
+      }
+    }
+    return changed ? success() : failure();
+  }
+};
+
+class AMDAIESinkIntoCorePass
+    : public impl::AMDAIESinkIntoCoreBase<AMDAIESinkIntoCorePass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<tensor::TensorDialect, linalg::LinalgDialect,
+                    AMDAIE::AMDAIEDialect, xilinx::AIE::AIEDialect>();
+  }
+
+  AMDAIESinkIntoCorePass() = default;
+  AMDAIESinkIntoCorePass(const AMDAIESinkIntoCorePass &pass) = default;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    patterns.insert<MoveAllDependenciesIntoOp<xilinx::AIE::CoreOp>>(context);
```

Comment:
What about operating on `AMDAIE::CoreOp` and moving it before `LowerToAIE`?

Reply:
See other comment :)